Drafts/1NNN-JSH.md

```diff
@@ -0,0 +1,143 @@
+| Implementation: | https://git.io/fpSUA                                            |
+| Status:         | Will be set by the DIP manager (e.g. "Approved" or "Rejected")  |
+
+## Abstract
+
+This DIP proposes adding a "string sequence literal" to D, primarily inspired
+by string interpolation, but also applicable to a wide variety of use cases.
+
+In a nutshell, this literal:
+
+````
+i"Hello, ${name}! You have logged on ${count} times."
+````
+
+is translated to the [compile-time sequence](https://dlang.org/articles/ctarguments.html):
+
+````D
+"Hello, ", name, "! You have logged on ", count, " times."
+````
+Note that the compiler does not perform any string interpolation; it merely
+segments the literal into a sequence of strings and expressions.  The intent is for
+further processing to be done in library code (see [rationale](#rationale) for a more detailed
+description of possible applications).
+
+### Reference
+
+- Exploration: https://github.com/marler8997/interpolated_strings
+- Example Library Solution: https://github.com/dlang/phobos/pull/6339/files
+- Implementation: https://github.com/dlang/dmd/pull/7988
+- https://forum.dlang.org/thread/khcmbtzhoouszkheqaob@forum.dlang.org
+- https://forum.dlang.org/thread/c2q7dt$67t$1@digitaldaemon.com
+- https://forum.dlang.org/thread/qpuxtedsiowayrhgyell@forum.dlang.org
+- https://forum.dlang.org/thread/ncwpezwlgeajdrigegee@forum.dlang.org
+- https://dlang.typeform.com/report/H1GTak/PY9NhHkcBFG0t6ig (#3 in "What language features do you miss?")
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Language vs Library Feature](#language-feature-vs-library-feature)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Sequence literals apply to a wide range of use cases. A few of these use cases are outlined below.
+
+#### String Interpolation
+One notable use for sequence literals is in string interpolation, which allows for more concise, readable, and maintainable code. For example:
+
+
+src/build.d:556:<br>
+`auto hostDMDURL = "http://downloads.dlang.org/releases/2.x/"~hostDMDVer~"/dmd."~hostDMDBase;`<br>
+Becomes:<br>
+`auto hostDMDURL = i"http://downloads.dlang.org/releases/2.x/$hostDMDVer/dmd.$hostDMDBase".text;`<br>
+And, with syntax highlighing:<br>
+![https://i.imgur.com/tXm6rBU.png](https://i.imgur.com/tXm6rBU.png)
+
+
+src/dmd/json.d:1058:<br>
+``s ~= prefix ~ "`" ~ enumName ~ "`";``<br>
+Becomes:<br>
+``s ~= i"prefix`$enumName`".text;``<br>
+With syntax highlighting:<br>
+![https://i.imgur.com/KTcOS0F.png](https://i.imgur.com/KTcOS0F.png)
+
+
+
+#### Database Queries
+
+`db.exec("UPDATE Foo SET a = ?, b = ?, c = ?, d = ? WHERE id = ?", aval, bval, cval, dval, id);`<br>
+Becomes:<br>
+`db.exec(i"UPDATE Foo SET a = $(aval), b = $(bval), c = $(cval), d = $(dval) WHERE id = $(id)");`
```

Comment:
Indeed, and actually, I would never, as a library author, write a library function that works with `i"..."` as it is defined right now because it would be so easy to do wrong.

What would stop you from writing `db.exec("UPDATE foo SET a = ", a)`? Well, nothing, and it would work fine, but just kinda coincidentally.

But what about `db.exec(i"UPDATE $(tablePrefix)table SET a = ?", 10)`. OK, it is fair to say "don't do that", but there's no way to tell, in the library, that they did!

It translates to `db.exec("UPDATE ",tablePrefix,"table SET a = ?", 10)`. At run time, you can issue a sql syntax error, but it is impossible to tell the difference at compile time.

I hate to be the downer, cuz I basically like this proposal, but I still think we should be making this thing give a new type. Instead of a naked tuple, do an anonymous struct. Change `text` and `writeln` etc. to recognize this new kind of struct, or at the usage site, use `.tupleof`.

`i"foo $(a)".tupleof.whatever is still an option... but then library authors who actually want to work with the details have them available. (and we should probably change text and writeln so those just work first anyway.)

Reply:
> I still think we should be making this thing give a new type. Instead of a naked tuple, do an anonymous struct

Amen. We need to be able to tell on the callee side if an interpolated sequence was passed vs just variadic args. And if it's a struct we can also have an alias member called `text` for std.conv.text, so we can actually get a string easily without having to keep adding a local import every time we use interpolated strings. Getting a string is a major use case, however nice parsing a sequence of elements is for efficiency.

Reply:
yeah, I can live with the convenience `.text` method. (i almost want to call it `.toString` to leverage existing things that call that name, but whatever, i don't care that much about the name).

I am slightly disgusted at having a magic struct created by the compiler include such a reference, but I think it works and the compiler already magically references std.math in syntax, so I can live with it. I'd just make its magic `.text` thing be a zero-arg template, so you don't pay for it if it isn't actually called.

I think I will write more about this in today's "this week in D"; I need to write that in a couple hours anyway and that gives me some content to dual-purpose since i missed most of dconf lol

I'll link it back here in a while. It might be a fairly different implementation though, since a magic struct cannot - I think anyway - be done with just a lexer hack. Even if you wrapped it in a call syntax to punt the details to the runtime library, you'd actually lose some functionality. (consider: `foo!i"$(some_alias)"`. If that is passed to `__d_is(some_alias)`, the aliasness is gone. On the other hand, `foo(i"$(a + 4)")` being passed to `__d_is!(a+4)` is liable to fail with "variable a is not accessible at compile time". So I don't think they *both* can work with a library function injected in the middle. Now, if I had to choose, I think the latter case is more important and I'd sacrifice the former case, but with the current PR, both work, and I think that is actually kinda cool. So I wanna preserve that. And a compiler-generated magic struct can do it.)

Anyway, I'll write more in a few hours.

Reply:
Yeah, since we have `alias this`, we can default to expanding the interpolated struct into a tuple while still giving access to methods to work with the interpolated data structure.

```D
writeln(i"1 + 2 is $(1+2)");

// don't need to import std.conv.text because .text is just a member of the
// interpolated string type
audo s = "1 + 2 is $(1 + 2)".text;
```

I'd like to know where @WalterBright  and @andralex stand on this.  Since they are the ones to decide what design will be accepted, I'd like to get their input on this particular aspect.  But I think I agree with you on this one.

Reply:
Well, it took me a lot of hours to write this, and the process of writing it had me change my mind about a few things so I hastily edited previous stuff, but regardless, read the whole thing to see where I ended up. Basically though, compiler calls a runtime function, function returns a struct with alias this. Lots of exciting potential with one thing beyond the tuple (aside from the type safety): keeping the original string as a template argument/enum member of the struct for further compile-time inspection.

Here's my blogged "DIP": http://dpldocs.info/this-week-in-d/Blog.Posted_2019_05_13.html

Reply:
Alot of details and good ideas.

I would add an `expand` member to your `__d_interpolated_string` type, just like the `Tuple` type has that expands the struct to a tuple, and then `alias this expand` which would allow:
```D
writeln(i"foo is $(foo)");
```

I would strongly prefer the alias this to default to a non-allocating tuple rather than `toString`.  I would be very wary of using interpolated strings otherwise.

Reply:
structs have a built-in property, `tupleof` which makes a separate `expand` function unnecessary (I actually don't know why the phobos one has a separate function at all, since .tupleof works just as well on it as any other struct)... you also cannot alias this to a tuple regardless - well, the compiler will allow it, but it will do nothing as no situation can ever trigger it, and moreover, writeln, being a template, won't trigger alias this anyway. templates work with the exact type passed to them.

Note that `writeln` would surely just learn how to handle the interpolated string anyway. It would be a ~3 line change to std.stdio to handle it gracefully internally with no allocations necessary regardless of `alias this`.

The alias this isn't there for writeln, it is just in case someone wants to do like `i"$(a)" ~ something else`, or `string s = i"$(a)";`. But since the former is a little bit silly and the latter is answered easily by saying `use the .toString method`, we could live without it.

(and note that there's another signature of toString the library supports, one that takes a sink delegate to avoid allocating. We could just implement that one too and handle a variety of cases. Lots of existing options to explore, and with the hybrid library approach, the compiler doesn't need to know or care.)

Reply:
I like Adam's proposal.  I think it adds more value to the basic interpolated string proposal, perhaps just enough additional value to tip the scale when the DIP is officially reviewed.

Reply:
My mistake, I forgot that `Tuple` doesn't automatically expand to a "tuple", that's kind of the whole reason why `Tuple` exists :)

In any case I do see alot of versatility in using the structure.  It's a very good idea and I'm on board with it.  It does make a few use cases harder by sometimes requiring the user to use `tupleof` and/or requiring a wrapper template, but a small cost compared to the benefits.  There will likely only be a handful of functions that need to accept interpolated strings.  Though, I would like to be able to do things like:
```
// contrived example, you wouldn't actually do it this way
writeln(i"foo is $(foo) and bar is ", bar, i" and baz is $(baz)");
```
This would mean I don't think we could just use a wrapper, but would need to add a `static if` inside `writeln` to handle interpolated strings to automatically expand them.

And yes, having a `toString(sink)` would also be good.

Thanks for taking the time to write this up, I like it alot.  I'll think about it and may update my implementation to it.

Reply:
One more comment on your interpolation proposal.  I think a big use case for interpolated strings is to make mixins much easier to write/maintain.  Using `q{}` string literals along with interpolated strings is a big win.  You mentioned in your article that you don't think they should be supported, and provided an example that you thought was ambiguous.  However, it's currently ambiguous in my implementation.

There's 2 ways you could go about string interpolation.  You could either parse the interpolated string at the same time you are you are parsing the string literal, or you could first parse the string literal and the interpolate it.  I opted for the second design.  With this model, it's easier to determine what will happen in cases like yours because the string is processed by each part one at a time.  So in your example:

```D
mixin(iq{
   string s = i"foo $(bar)";
});
```

1. The token string is parsed.  Token strings don't have escapes so it stays the same.
2. Then the interpolater runs.  It ignores everything except when it sees a '$' character.  The rule is made intentionally simple so a developer will be able to determine what will happen.  The the string above will become:

```D
mixin(__d_interpolatedString(`
   string s = i"foo `, bar, `";
`));
```

Then mixin will get involved, of which interpolatedString support should be added, and it will convert `bar` to a string, let's suppose it's string representation is `"barvalue"`.
```D
mixin(`
   string s = i"foo barvalue";
`));
```
```D
   string s = i"foo barvalue";
```

Then the interpolator runs again, and we just get:
```D
     string s = __d_interpolatedString("foo barvalue");
```

Now if you wanted the other behavior where the `$(bar)` expression is expanded later, then you would escape the `$` with `$$`:

```D
mixin(iq{
   string s = i"foo $$(bar)";
});
```
Will become:
```D
   string s = i"foo $(bar)";
```
Which becomes:
```D
     string s = __d_interpolatedString("foo ", bar);
```

So by default, any `$(...)` expressions are always expanded by the initial interpolation, and if you want to delay expansion then you can escape the dollar.

Reply:
By the way, double quotes strings have the same issue.  Your example doesn't only apply to token strings:
```D
mixin("
   string s = i\"foo $(bar)\";
");
```

Reply:
Well, I think it is trivial with double-quoted strings because it is all syntax highlighted as a big string :) with token strings, the D lexer runs inside too so you might think it is different.

But that said, your reasoning is solid and simple to understand, so I could go with that too. (just gotta be sure we mention all these things - and explicitly mention what we decide to exclude - to get past the Walter barrier :) )

And also I guess mixin should probably be able to handle the interpolated thing too. The `alias this` could handle that, so could an explicit call to `toString` (or `text`), but we should prolly document it regardless.

PS this is a kinda weird place to be having this conversation, as comments on a random line of example code :P but whatever.

Reply:
Using `alias this` with `toString` would mean that using mixins with interpolated strings would require phobos.
```D
enum SomeValue = 1234;
mixin("enum AnotherValue = $(SomeValue + 1)");
```

Seems wrong to me....I'll have to think about this one.